pkg/gig/service.go

```diff
@@ -0,0 +1,302 @@
+package gig
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/cortezaproject/corteza-server/pkg/id"
+)
+
+var (
+	// wrapper around time.Now() that will aid service testing
+	now = func() *time.Time {
+		c := time.Now().Round(time.Second)
+		return &c
+	}
+
+	// wrapper around id.Next() that will aid service testing
+	nextID = func() uint64 {
+		return id.Next()
+	}
+
+	gigStore = make(map[uint64]*Gig)
+)
+
+type (
+	Service interface {
+		Create(context.Context, Worker) (*Gig, error)
+		Read(context.Context, uint64) (*Gig, error)
+
+		Output(context.Context, uint64) (SourceSet, error)
+		Status(context.Context, uint64) (WorkerStatus, error)
+
+		Prepare(context.Context, uint64) error
+		Exec(context.Context, uint64) error
+
+		AddSources(ctx context.Context, gigID uint64, sources SourceSet, decoders ...DecodeTask) (g *Gig, err error)
+		RemoveSources(ctx context.Context, gigID uint64) (g *Gig, err error)
+		SetDecoders(ctx context.Context, gigID uint64, decoders ...DecodeTask) (*Gig, error)
+		SetPreprocessors(ctx context.Context, gigID uint64, preprocessors ...PreprocessTask) (*Gig, error)
+		SetPostprocessors(ctx context.Context, gigID uint64, postprocessors ...PostprocessTask) (*Gig, error)
+		Cleanup(ctx context.Context, gigID uint64) (*Gig, error)
+	}
+
+	service struct {
+		// @todo
+		opt map[string]interface{}
+	}
+)
+
+// Initialize a new gig service
+func NewService(opt map[string]interface{}) *service {
+	return &service{
+		opt: opt,
+	}
+}
+
+// Create initializes a new gig
+func (svc *service) Create(ctx context.Context, w Worker) (g *Gig, err error) {
+	return updateGig(ctx, newGig(w))
+}
+
+// Read returns the requested gig
+func (svc *service) Read(ctx context.Context, id uint64) (*Gig, error) {
+	return getGig(ctx, id)
+}
+
+func (svc *service) AddSources(ctx context.Context, gigID uint64, sources SourceSet, decoders ...DecodeTask) (g *Gig, err error) {
```

Comment:
Why do you want to keep these gig-modifying functions on the service and not on the gig struct (or as a standalone functions)? Where do we benefit if we drive all this changes through gig service?

Reply:
Discussed